bgpd/bgp_mplsvpn.c

```diff
@@ -1336,6 +1336,225 @@ leak_update(struct bgp *to_bgp, struct bgp_dest *bn,
+	 */
+	if (bnc->nexthop->type != NEXTHOP_TYPE_IFINDEX) {
+		if (bnc->nexthop->type == NEXTHOP_TYPE_IPV4 ||
+		    bnc->nexthop->type == NEXTHOP_TYPE_IPV4_IFINDEX) {
+			nh_from_bnc.family = AF_INET;
+			nh_from_bnc.prefixlen = IPV4_MAX_BITLEN;
+			nh_from_bnc.u.prefix4 = bnc->nexthop->gate.ipv4;
+		} else {
+			nh_from_bnc.family = AF_INET6;
+			nh_from_bnc.prefixlen = IPV6_MAX_BITLEN;
+			memcpy(&nh_from_bnc.u.prefix6, &bnc->nexthop->gate.ipv6,
+			       sizeof(struct in6_addr));
+		}
+		nh_to_use_ptr = &nh_from_bnc;
+	} else
+		/* the nexthop is in bnc->prefix */
+		nh_to_use_ptr = &bnc->prefix;
+
+	/* at least one bnc is connected.
+	 * first, let us allocate a cache context
+	 */
+	tree = &from_bgp->mpls_labels_per_nexthop[family2afi(
+		nh_to_use_ptr->family)];
+	blnc = bgp_label_per_nexthop_find(tree, nh_to_use_ptr);
+	if (!blnc) {
+		blnc = bgp_label_per_nexthop_new(tree, nh_to_use_ptr);
+		blnc->to_bgp = to_bgp;
+	}
+
+	if (pi->label_nexthop_cache != blnc) {
+		/* Unlink from existing nexthop cache, if any.
+		 * This will also free the nexthop label cache entry
+		 */
+		bgp_mplsvpn_path_nh_label_unlink(pi);
+		if (blnc) {
+			/* updates NHT pi list reference */
+			LIST_INSERT_HEAD(&(blnc->paths), pi, label_nh_thread);
+			pi->label_nexthop_cache = blnc;
+			pi->label_nexthop_cache->path_count++;
+		}
+	}
+
+	if (blnc->label == MPLS_INVALID_LABEL)
+		bgp_lp_get(LP_TYPE_NEXTHOP, blnc,
+			   bgp_mplsvpn_get_label_per_nexthop_cb);
+	return blnc->label;
+}
+
+static mpls_label_t vpn_leak_from_vrf_get_label_for_path_vrf(
+	afi_t afi, safi_t safi, struct bgp_path_info *path_vrf,
+	struct bgp *from_bgp, struct bgp *to_bgp)
+{
+	mpls_label_t label_val;
+	bool attempt_policy_per_nexthop = false;
+	struct bgp *bgp_nexthop = NULL;
+	afi_t nh_afi;
+
+	/* only bgp updates that have a nexthop with a label can be sent */
+	if (CHECK_FLAG(from_bgp->vpn_policy[afi].flags,
+		       BGP_VPN_POLICY_TOVPN_LABEL_PER_NEXTHOP)) {
+		if (afi == AFI_IP) {
+			if (CHECK_FLAG(path_vrf->attr->flag,
+				       ATTR_FLAG_BIT(BGP_ATTR_NEXT_HOP)) &&
+			    path_vrf->attr->nexthop.s_addr != INADDR_ANY &&
+			    ipv4_unicast_valid(&path_vrf->attr->nexthop))
+				attempt_policy_per_nexthop = true;
+			else if ((path_vrf->attr->mp_nexthop_len ==
+				  BGP_ATTR_NHLEN_IPV4) &&
+				 path_vrf->attr->mp_nexthop_global_in.s_addr !=
+					 INADDR_ANY &&
+				 ipv4_unicast_valid(
+					 &path_vrf->attr->mp_nexthop_global_in))
+				attempt_policy_per_nexthop = true;
+		} else if (afi == AFI_IP6 &&
+			   (path_vrf->attr->mp_nexthop_len ==
+				    BGP_ATTR_NHLEN_IPV6_GLOBAL ||
+			    path_vrf->attr->mp_nexthop_len ==
+				    BGP_ATTR_NHLEN_IPV6_GLOBAL_AND_LL) &&
+			   !IN6_IS_ADDR_UNSPECIFIED(
+				   &path_vrf->attr->mp_nexthop_global) &&
+			   !IN6_IS_ADDR_LOOPBACK(
+				   &path_vrf->attr->mp_nexthop_global) &&
+			   !IN6_IS_ADDR_MULTICAST(
+				   &path_vrf->attr->mp_nexthop_global))
+			attempt_policy_per_nexthop = true;
+	}
+	if (attempt_policy_per_nexthop) {
```

Comment:
Move the code from that block to `bgp_mplsvpn_get_label_for_path_vrf`
and replace `attempt_policy_per_nexthop = true;` by a call to `bgp_mplsvpn_get_label_for_path_vrf`

Reply:
some code has been moved to the ```bgp_mplsvpn_get_label_for_path_vrf``` function. an extra parameter `safi` has been added.